### PR TITLE
cmd/derper,derp: add --rate-config file with SIGHUP reload

### PR DESCRIPTION
**Pull request by mikeodr**
Add a --rate-config flag pointing to a JSON file for per-client receive
rate limits (bytes/sec and burst bytes). The config is reloaded on SIGHUP,
updating all existing client connections live. The --per-client-rate-limit
and --per-client-rate-burst flags are removed in favor of the config file.

In derpserver, rate limiting uses an atomic.Pointer[xrate.Limiter] per
client: nil when unlimited or mesh (zero overhead), non-nil when
rate-limited.

Document that clientSet.activeClient Store operations require Server.mu.

Updates tailscale/corp#38509

### CODE REVIEW

File: cmd/derper/derper.go [hunk 1]
```diff
@@ -89,6 +89,7 @@ var (
 
 	perClientRateLimit = flag.Uint("per-client-rate-limit", 0, "per-client receive rate limit in bytes/sec; 0 means unlimited. Mesh peers are exempt.")
 	perClientRateBurst = flag.Uint("per-client-rate-burst", 0, "per-client receive rate burst in bytes; 0 defaults to 2x the rate limit (only relevant when using nonzero --per-client-rate-limit)")
+	rateConfigPath     = flag.String("rate-config", "", "path to JSON rate limit config file; reloaded on SIGHUP. Overrides --per-client-rate-limit and --per-client-rate-burst flags when set.")
```
Comment (jwhited) [inline cmd/derper/derper.go]: We also have --config, which right now just contains the private key. Perhaps this can sit alongside? Thoughts?

Comment (mikeodr) [reply cmd/derper/derper.go]: Yeah I specifically put this elsewhere as I was concerned about it getting changed or overwritten if other tools are updating the rate limit. 

I'm open to changing it.

Comment (jwhited) [reply cmd/derper/derper.go]: We decided separation from private key is preferred, especially considering deployment tooling having to read first if they were combined.

Comment (jwhited) [inline cmd/derper/derper.go]: If we're doing a config file we might as well remove the flags. No one is using them yet.

File: derp/derpserver/derpserver.go [hunk 2]
```diff
@@ -518,14 +518,110 @@ func (s *Server) SetTCPWriteTimeout(d time.Duration) {
...
-// than [derp.MaxPacketSize]. This ensures at least a full packet can
-// be received in a burst, even if the rate limit is low.
-func (s *Server) SetPerClientRateLimit(bytesPerSec, burst uint) {
+// RateConfig is a JSON-serializable configuration for per-client rate limits.
+// Values are specified in megabits per second (Mbps). Use [MbpsToBytes] to
+// convert to bytes per second for use with [Server.UpdatePerClientRateLimit].
+// A zero value for either field means unlimited.
+type RateConfig struct {
+	PerClientRateLimitMbps float64 `json:",omitzero"`
+	PerClientRateBurstMbps float64 `json:",omitzero"`
```
Comment (jwhited) [inline derp/derpserver/derpserver.go]: Do we need fractions of a megabit per second? uint64?

Comment (jwhited) [reply derp/derpserver/derpserver.go]: Decision: let's express these in bytes and convert elsewhere. Keep it natural within derp server logic.

File: derp/derpserver/derpserver.go [hunk 3]
```diff
@@ -518,14 +518,110 @@ func (s *Server) SetTCPWriteTimeout(d time.Duration) {
...
+// BytesPerSec returns the rate limit and burst in bytes per second.
+// If burst is 0 and rate > 0, burst defaults to 2x the rate.
+func (rc RateConfig) BytesPerSec() (rateLimit, rateBurst uint) {
+	rateLimit = MbpsToBytes(rc.PerClientRateLimitMbps)
+	rateBurst = MbpsToBytes(rc.PerClientRateBurstMbps)
+	if rateBurst == 0 && rateLimit > 0 {
+		rateBurst = rateLimit * 2
+	}
+	return rateLimit, rateBurst
+}
```
Comment (jwhited) [inline derp/derpserver/derpserver.go]: These can be removed once we express the config in bytes.

File: derp/derpserver/derpserver.go [hunk 4]
```diff
@@ -518,14 +518,110 @@ func (s *Server) SetTCPWriteTimeout(d time.Duration) {
...
+	if path == "" {
+		return RateConfig{}, errors.New("rate config path is empty")
+	}
+	b, err := os.ReadFile(path)
+	if err != nil {
+		return RateConfig{}, fmt.Errorf("reading rate config: %w", err)
+	}
+	var rc RateConfig
+	dec := json.NewDecoder(bytes.NewReader(b))
+	dec.DisallowUnknownFields()
```
Comment (jwhited) [inline derp/derpserver/derpserver.go]: I like the idea here, but we may want flexibility to have unknown fields around evolution of the config.

File: derp/derpserver/derpserver.go [hunk 5]
```diff
@@ -518,14 +518,110 @@ func (s *Server) SetTCPWriteTimeout(d time.Duration) {
...
+	} else if v > 0 && MbpsToBytes(v) == 0 {
+		return fmt.Errorf("rate config: PerClientRateLimitMbps value %v is too small (converts to 0 bytes/sec)", v)
+	}
+	if v := rc.PerClientRateBurstMbps; v < 0 || v > maxMbps {
+		return fmt.Errorf("rate config: PerClientRateBurstMbps must be between 0 and %.0f, got %v", maxMbps, v)
+	} else if v > 0 && MbpsToBytes(v) == 0 {
+		return fmt.Errorf("rate config: PerClientRateBurstMbps value %v is too small (converts to 0 bytes/sec)", v)
+	}
+	return nil
+}
```
Comment (jwhited) [inline derp/derpserver/derpserver.go]: Moving away from float64's in the config file probably removes the need for this.

File: derp/derpserver/derpserver.go [hunk 6]
```diff
@@ -518,14 +518,110 @@ func (s *Server) SetTCPWriteTimeout(d time.Duration) {
...
+
+// LoadAndApplyRateConfig reads a [RateConfig] from the file at path and
+// applies it to the server via [Server.UpdatePerClientRateLimit].
+func (s *Server) LoadAndApplyRateConfig(path string) error {
+	rc, err := LoadRateConfig(path)
+	if err != nil {
+		return err
+	}
+	rl, rb := rc.BytesPerSec()
```
Comment (jwhited) [inline derp/derpserver/derpserver.go]: This can be removed.

File: derp/derpserver/derpserver.go [hunk 7]
```diff
@@ -518,14 +518,110 @@ func (s *Server) SetTCPWriteTimeout(d time.Duration) {
...
+// The burst is at least [derp.MaxPacketSize], ensuring at least a full packet
+// can be received in a burst even if the rate limit is low. If bytesPerSec is
+// 0, rate limiting is set to infinity. Mesh peers are always exempt from rate
+// limiting.
+func (s *Server) UpdatePerClientRateLimit(bytesPerSec, burst uint) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
 	s.perClientRecvBytesPerSec = bytesPerSec
-	s.perClientRecvBurst = max(burst, derp.MaxPacketSize)
+	s.perClientRecvBurst = min(max(burst, derp.MaxPacketSize), math.MaxInt)
```
Comment (jwhited) [inline derp/derpserver/derpserver.go]: This can go back to original expression since we will use uint64's in the config.

File: derp/derpserver/derpserver.go [hunk 8]
```diff
@@ -1714,9 +1836,11 @@ type sclient struct {
...
 
-	// recvLim is the per-connection receive rate limiter. If non-nil,
-	// the server calls WaitN per received DERP frame in order to
-	// apply TCP backpressure and throttle the sender.
+	// recvLim is the per-connection receive rate limiter, always
+	// initialized by [Server.initClientRateLimitLocked]. When rate limiting
+	// is disabled or the client is a mesh peer, the limit is set to
+	// [xrate.Inf]. The server calls WaitN per received DERP frame
+	// in order to apply TCP backpressure and throttle the sender.
 	recvLim *xrate.Limiter
```
Comment (jwhited) [inline derp/derpserver/derpserver.go]: We can also make this an `atomic.Pointer` to reduce overheads for mesh and environments that do not rate limit. `rate.Limiter.WaitN` does a bunch of locking/unlocking and channel operations, even when the limit is infinity.

File: derp/derpserver/derpserver.go [hunk 9]
```diff
@@ -518,14 +518,110 @@ func (s *Server) SetTCPWriteTimeout(d time.Duration) {
...
+		// are in cs.dup.set, so we only need to iterate cs.dup.set.
+		if cs.dup != nil {
+			for c := range cs.dup.set {
+				c.setRateLimitLocked(bytesPerSec, burstInt)
+			}
+		} else if c := cs.activeClient.Load(); c != nil {
+			c.setRateLimitLocked(bytesPerSec, burstInt)
+		}
+	}
 }
```
Comment (jwhited) [inline derp/derpserver/derpserver.go:573]: This looks sounds to me at first glance, assuming that `clientSet.activeClient` Store calls also hold `Server.mu`. We don't have explicit docs saying this is the intent, but the current store calls are all doing this with `Server.mu` held. Perhaps we should update docs?

File: derp/derpserver/derpserver_test.go [hunk 10]
```diff
@@ -1040,37 +1042,504 @@ func TestPerClientRateLimit(t *testing.T) {
...
+		if got := c.recvLim.Limit(); got != rate.Limit(testRate) {
+			t.Errorf("client %d: rate limit = %v; want %d", i, got, testRate)
+		}
+		if got := c.recvLim.Burst(); got != int(testBurst) {
+			t.Errorf("client %d: burst = %v; want %d", i, got, testBurst)
+		}
+	}
+}
+
+func TestMbpsToBytes(t *testing.T) {
```
Comment (jwhited) [inline derp/derpserver/derpserver_test.go]: this can go away

File: derp/derpserver/derpserver_test.go [hunk 11]
```diff
@@ -1040,37 +1042,504 @@ func TestPerClientRateLimit(t *testing.T) {
...
+		{0.000001, 0},
+	}
+	for _, tt := range tests {
+		if got := MbpsToBytes(tt.mbps); got != tt.want {
+			t.Errorf("MbpsToBytes(%v) = %d; want %d", tt.mbps, got, tt.want)
+		}
+	}
+}
+
+func TestRateConfigBytesPerSec(t *testing.T) {
```
Comment (jwhited) [inline derp/derpserver/derpserver_test.go]: This can be updated for natural bytes.

File: derp/derpserver/derpserver_test.go [hunk 12]
```diff
@@ -1040,37 +1042,504 @@ func TestPerClientRateLimit(t *testing.T) {
...
+	t.Run("negative_burst_rejected", func(t *testing.T) {
+		f := filepath.Join(t.TempDir(), "rate.json")
+		if err := os.WriteFile(f, []byte(`{"PerClientRateBurstMbps": -1}`), 0644); err != nil {
+			t.Fatal(err)
+		}
+		_, err := LoadRateConfig(f)
+		if err == nil {
+			t.Fatal("expected error for negative burst")
+		}
+	})
```
Comment (jwhited) [inline derp/derpserver/derpserver_test.go]: This can go away when config is expressed as uint64

File: derp/derpserver/derpserver.go [hunk 13]
```diff
@@ -518,14 +518,65 @@ func (s *Server) SetTCPWriteTimeout(d time.Duration) {
...
+		// When cs.dup is non-nil, all clients (including the active one)
+		// are in cs.dup.set, so we only need to iterate cs.dup.set.
+		if cs.dup != nil {
+			for c := range cs.dup.set {
+				c.setRateLimitLocked(bytesPerSec, burstInt)
+			}
+		} else if c := cs.activeClient.Load(); c != nil {
+			c.setRateLimitLocked(bytesPerSec, burstInt)
+		}
+	}
```
Comment (jwhited) [inline derp/derpserver/derpserver.go:572]: We should probably use `clientSet.ForeachClient` instead, which does the same thing and simplifies this code.

File: derp/derpserver/derpserver.go [hunk 14]
```diff
@@ -239,10 +239,10 @@ type clientSet struct {
...
 	// if there are no connections or the connection is disabled.
 	//
-	// A pointer to a clientSet can be held by peers for long periods of time
-	// without holding Server.mu to avoid mutex contention on Server.mu, only
-	// re-acquiring the mutex and checking the clients map if activeClient is
-	// nil.
+	// Stores to activeClient must hold Server.mu. Loads may be done without
+	// holding Server.mu — a pointer to a clientSet can be held by peers for
+	// long periods of time to avoid mutex contention, only re-acquiring the
+	// mutex and checking the clients map if activeClient is nil.
```
Comment (jwhited) [inline derp/derpserver/derpserver.go]: I'd revert this when updating to use `clientSet.ForeachClient`, which we already have docs for on `clientSet`:
> // All methods should only be called while holding Server.mu.

File: derp/derpserver/derpserver.go [hunk 15]
```diff
@@ -518,14 +518,65 @@ func (s *Server) SetTCPWriteTimeout(d time.Duration) {
...
 }
 
-// SetPerClientRateLimit sets the per-client receive rate limit in bytes per
-// second and the burst size in bytes. Mesh peers are exempt from this limit.
-// The burst is at least [derp.MaxPacketSize], or burst, if burst is greater
-// than [derp.MaxPacketSize]. This ensures at least a full packet can
-// be received in a burst, even if the rate limit is low.
-func (s *Server) SetPerClientRateLimit(bytesPerSec, burst uint) {
+// RateConfig is a JSON-serializable configuration for per-client rate limits.
+// Values are in bytes. A zero value for either field means unlimited.
```
Comment (jwhited) [inline derp/derpserver/derpserver.go]: A zero value for `PerClientRateBurstBytes` can't really represent unlimited, can it? The bucket has a finite depth/capacity.

Comment (jwhited) [reply derp/derpserver/derpserver.go]: Or is this saying, if either field is 0, then no rate limit applies?

File: derp/derpserver/derpserver.go [hunk 16]
```diff
@@ -518,14 +518,65 @@ func (s *Server) SetTCPWriteTimeout(d time.Duration) {
...
+type RateConfig struct {
+	PerClientRateLimitBytesPerSec uint64 `json:",omitzero"`
+	PerClientRateBurstBytes       uint64 `json:",omitzero"`
```
Comment (jwhited) [inline derp/derpserver/derpserver.go]: maybe something like (comments may need multiple lines):

```suggestion
// Values are in bytes.
type RateConfig struct {
    // PerClientRateLimitBytesPerSec represents the per-client rate limit in bytes per second. A zero value disables rate-limiting.
	PerClientRateLimitBytesPerSec uint64 `json:",omitzero"`
	// PerClientRateBurstBytes represents the per-client token bucket depth, or burst, in bytes. Any value lower than [derp.MaxPacketSize] will be increased to [derp.MaxPacketSize] before application.
	PerClientRateBurstBytes       uint64 `json:",omitzero"`
```

File: derp/derpserver/derpserver.go [hunk 17]
```diff
@@ -518,14 +518,65 @@ func (s *Server) SetTCPWriteTimeout(d time.Duration) {
...
+	return nil
+}
+
+// UpdatePerClientRateLimit sets the per-client receive rate limit in bytes per
+// second and the burst size in bytes, updating all existing client connections.
+// The burst is at least [derp.MaxPacketSize], ensuring at least a full packet
+// can be received in a burst even if the rate limit is low. If bytesPerSec is
+// 0, rate limiting is set to infinity. Mesh peers are always exempt from rate
+// limiting.
+func (s *Server) UpdatePerClientRateLimit(bytesPerSec, burst uint) {
```
Comment (jwhited) [inline derp/derpserver/derpserver.go]: We have `uint` here and in the `Server` fields, `uint64` in the JSON struct, and `int` in the x/rate APIs, and type conversion in multiple places.

Let's simplify to `uint64` everywhere, and just convert around the x/rate API.

File: derp/derpserver/derpserver.go [hunk 18]
```diff
@@ -1267,19 +1317,51 @@ func (c *sclient) handleFrameSendPacket(_ derp.FrameType, fl uint32) error {
...
+// s.mu must be held.
+func (s *Server) initClientRateLimitLocked(c *sclient) {
+	c.setRateLimitLocked(s.perClientRecvBytesPerSec, int(max(s.perClientRecvBurst, derp.MaxPacketSize)))
+}
+
+// setRateLimitLocked updates the per-client receive rate limiter.
+// When bytesPerSec is 0 or the client is a mesh peer, the limiter is
+// set to nil so that [sclient.rateLimit] is a no-op.
+//
+// s.mu must be held.
```
Comment (jwhited) [inline derp/derpserver/derpserver.go]:
```suggestion
// Server.mu must be held.
```

File: derp/derpserver/derpserver.go [hunk 19]
```diff
@@ -1267,19 +1317,51 @@ func (c *sclient) handleFrameSendPacket(_ derp.FrameType, fl uint32) error {
...
+func (c *sclient) setRateLimitLocked(bytesPerSec uint, burst int) {
```
Comment (jwhited) [inline derp/derpserver/derpserver.go]: should probably be named `setRateLimitServerLocked`, and docs updated

File: derp/derpserver/derpserver.go (hunk 18 above)
Comment (jwhited) [reply derp/derpserver/derpserver.go]: Also, it's nonobvious why `Server.mu` must be held, which is worth explaining a bit more. The function body only deals with atomics.

File: derp/derpserver/derpserver_test.go [hunk 20]
```diff
@@ -1040,37 +1042,292 @@ func TestPerClientRateLimit(t *testing.T) {
...
+
+		clientKey := key.NewNode().Public()
+		c := &sclient{key: clientKey, s: s, logf: logger.Discard}
+		cs := &clientSet{}
+		cs.activeClient.Store(c)
+		s.mu.Lock()
+		s.clients[clientKey] = cs
+		s.mu.Unlock()
+
+		f := writeConfig(t, `{"PerClientRateLimitBytesPerSec": 1250000, "PerClientRateBurstBytes": 2500000}`)
```
Comment (jwhited) [inline derp/derpserver/derpserver_test.go:1274]: We should also test a zero `PerClientRateBurstBytes` turns into `derp.MaxPacketSize`. Not sure if that belongs in this test, but I don't see it in any test, currently.

File: derp/derpserver/derpserver.go (hunk 18 above)
Comment (mikeodr) [reply derp/derpserver/derpserver.go]: I think the previous impl needed this to be locked, but it no longer does.

File: derp/derpserver/derpserver.go [hunk 21]
```diff
@@ -1267,19 +1317,51 @@ func (c *sclient) handleFrameSendPacket(_ derp.FrameType, fl uint32) error {
...
 
-// rateLimit applies the per-client receive rate limit, if configured.
+// initClientRateLimitLocked initializes the receive rate limiter for c using
+// the server's current configuration. The limiter is always created so that
+// [Server.UpdatePerClientRateLimit] can safely update it later without
+// nil checks. Called from [Server.registerClient] so that initialization
+// and registration are atomic under s.mu.
+//
+// s.mu must be held.
+func (s *Server) initClientRateLimitLocked(c *sclient) {
```
Comment (jwhited) [inline derp/derpserver/derpserver.go]: This is a one-line function that is only called from `Server.RegisterClient`. Can it move there?

File: derp/derpserver/derpserver.go (hunk 15 above)
Comment (mikeodr) [reply derp/derpserver/derpserver.go]: Yeah this is wrong, only the PerClientRateLimitBytesPerSec applies to the comment.
The other comments you suggested are more correct.

File: derp/derpserver/derpserver.go (hunk 14 above)
Comment (mikeodr) [reply derp/derpserver/derpserver.go]: This is already documented above, yeah reverting.

File: derp/derpserver/derpserver.go [hunk 22]
```diff
@@ -1714,10 +1796,13 @@ type sclient struct {
...
-	// the server calls WaitN per received DERP frame in order to
-	// apply TCP backpressure and throttle the sender.
-	recvLim *xrate.Limiter
+	// recvLim is the per-connection receive rate limiter. When rate
+	// limiting is enabled for a non-mesh client, it points to an
+	// [xrate.Limiter]. When rate limiting is disabled or the client is a
+	// mesh peer, it is nil and [sclient.rateLimit] is a no-op.
+	// Updated atomically by [sclient.setRateLimitLocked] so that
+	// [sclient.rateLimit] can load it without holding Server.mu.
+	recvLim atomic.Pointer[xrate.Limiter]
```
Comment (jwhited) [inline derp/derpserver/derpserver.go:1787]: We should TODO about mono time optimizations. In fact, we already use `tstime/rate.Limiter` in the DERP server elsewhere, but the API surface is missing some behaviors we need for `recvLim`.

File: derp/derpserver/derpserver.go (hunk 18 above)
Comment (jwhited) [reply derp/derpserver/derpserver.go]: `sClient` has no sync primitives guarding the order of operations contained: the load, mutate, then store requires sync if this method can be called concurrently. That should be documented at minimum.

This method is not currently called concurrently, but holding `Server.mu` is a safeguard against future concurrent calls if the code changes shape, which is probably how it ended up this way.

File: derp/derpserver/derpserver.go [hunk 23]
```diff
@@ -518,14 +514,64 @@ func (s *Server) SetTCPWriteTimeout(d time.Duration) {
...
+	defer s.mu.Unlock()
 	s.perClientRecvBytesPerSec = bytesPerSec
 	s.perClientRecvBurst = max(burst, derp.MaxPacketSize)
+	burstInt := s.perClientRecvBurst
+
+	for _, cs := range s.clients {
+		cs.ForeachClient(func(s *sclient) {
+			s.setRateLimit(bytesPerSec, burstInt)
+		})
+	}
```
Comment (jwhited) [inline derp/derpserver/derpserver.go]:
```suggestion
	for _, cs := range s.clients {
		cs.ForeachClient(func(c *sclient) {
			c.setRateLimit(s.perClientRecvBytesPerSec, s.perClientRecvBurst)
		})
	}
```

File: derp/derpserver/derpserver.go (hunk 21 above)
Comment (jwhited) [reply derp/derpserver/derpserver.go]: It got a doc update, but didn't move :)

File: derp/derpserver/derpserver.go [hunk 24]
```diff
@@ -690,6 +734,8 @@ func (s *Server) registerClient(c *sclient) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
+	c.setRateLimit(s.perClientRecvBytesPerSec, max(s.perClientRecvBurst, derp.MaxPacketSize))
```
Comment (jwhited) [inline derp/derpserver/derpserver.go]: The only mutation of `Server.PerClientRecvBurst` does the same max(). If we change max() logic in the future it's better to only have it in one place.

**Review by jwhited — APPROVED** [review]
